# Dataset Inventory — Financial Manager AI

## Objective

This notebook performs an initial inventory of the datasets considered for the Financial Manager AI project.

The purpose of this stage is to understand the structure, quality, granularity, and potential applicability of each dataset before performing data cleaning, transformation, modeling, or database integration.

The following aspects will be evaluated:

- dataset structure;
- number of records and attributes;
- data types;
- missing values;
- attribute cardinality;
- possible identifiers;
- dimensions and measures;
- temporal coverage;
- record granularity;
- potential relationships between attributes;
- applicability to the Financial Manager AI modules.

No permanent data cleaning or transformation will be performed at this stage.

## 1. Imports

Import the libraries required for the initial dataset inspection.

In [1]:
import pandas as pd

## 2. Dataset Loading

The dataset analyzed in this section is the **Receiver General Accounting Transactions** dataset.

The original raw file is loaded without modifying its values, columns, or structure.    

In [2]:
df = pd.read_csv(
    "../data/raw/Receiver General Accounting Transactions/tcrg-rgat.csv"
)

### Dataset Metadata

- **Dataset:** Receiver General Accounting Transactions
- **Local file:** `tcrg-rgat.csv` (CSV), preserved unchanged.
- **Authoritative publisher:** Public Services and Procurement Canada (PSPC), acting as Receiver General for Canada; [Government of Canada catalogue record](https://open.canada.ca/data/en/dataset/fa79f13c-2796-41a9-a9cd-28650ebe0f06), identifier `fa79f13c-2796-41a9-a9cd-28650ebe0f06`.
- **Redistribution/acquisition source:** Kaggle, as recorded in the original inventory. The exact Kaggle listing, uploader, version, and download date have not been established; this attribution is retained, not independently verified.
- **Official resources:** the catalogue links a [current-fiscal-year CSV](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/tcrg-rgat/tcrg-rgat.csv) and a [previous-years CSV](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/tcrg-rgat/tcrg-rgat-arch.csv). Matching the local filename identifies the resource naming convention, not the local extract's version or byte-for-byte provenance. Neither official CSV was downloaded for this review.
- **Documentation / Data Dictionary:** [Banking and Cash Management XML dictionary](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/SGBT-BCMS-dd.xml), explicitly linked by the dataset catalogue; [dataset-specific supporting documentation](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/tcrg-rgat/soutien-support-eng.html).
- **Documentation review date:** 2026-09-26. The supporting page displays a modification date of 2024-04-26; these are current reference documents, not a verified historical documentation snapshot for the local extract.

The official catalogue describes automated and manually adjusted accounting entries received before further validation for the Receiver General general ledger. This identifies the dataset's origin and processing context; it does not establish that any particular local value is erroneous. The remaining provenance gap is the exact chain from an official extract to the recorded Kaggle redistribution and this local file. [Source: official catalogue](https://open.canada.ca/data/en/dataset/fa79f13c-2796-41a9-a9cd-28650ebe0f06).


## 3. General Dataset Structure

This section provides an initial overview of the dataset, including its number of rows, number of columns, column names, data types, memory usage, and non-null values.

This information helps identify the overall structure of the dataset before performing more detailed analysis.

In [3]:
df.shape

(9282, 16)

In [4]:
df.columns.tolist()

['Accounting-Control-Number-Numéro-contrôle-comptable',
 'Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal',
 'Fiscal-Year-Année-Fiscale',
 'Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal',
 "Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable",
 'DepartmentNumber-Numéro-de-Ministère',
 'Fiscal-Month-Mois-Fiscal',
 "Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal",
 'Financial-Coding-Block-Bloc-de-codage-financier',
 'Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire',
 'Credit/Debit-Code-Code-Crédit/Débit',
 "Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal",
 'Control-Data-Type-Type-de-données-de-contrôle',
 'Accounting-Summary-Transaction-Line-Count-Sommaire-comptable-transaction-compte-ligne',
 'GBS-Control-Number-Numéro-de-contrôle-du-SBG',
 'General-Ledger-Account-Code-Code-du-compte-du-grand-livre-général']

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9282 entries, 0 to 9281
Data columns (total 16 columns):
 #   Column                                                                                 Non-Null Count  Dtype  
---  ------                                                                                 --------------  -----  
 0   Accounting-Control-Number-Numéro-contrôle-comptable                                    9282 non-null   int64  
 1   Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                             9256 non-null   str    
 2   Fiscal-Year-Année-Fiscale                                                              9282 non-null   str    
 3   Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                       178 non-null    float64
 4   Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                           9282 non-null   str    
 5   DepartmentNumber-Numéro-de-Ministère                                                   

### Initial Observations

The dataset contains 9,282 records and 16 attributes.

The columns include integer, floating-point, and string data types.

Some attributes contain substantial amounts of missing data, while several core accounting fields appear to be complete.

## 4. Data Sample and Record Granularity

A small sample of records is inspected to better understand the meaning of the attributes and determine what a single row represents.

The dataset is transposed for visualization purposes because the original column names are long and difficult to inspect horizontally.

In [6]:
print(df.head().T.to_string())

                                                                                                       0                  1                  2                  3                  4
Accounting-Control-Number-Numéro-contrôle-comptable                                         202309300033       202309300034       202309300035       202309300036       202309300037
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                                         JVEFI              JVEFI              JVEFI              JVEFI              JVEFI
Fiscal-Year-Année-Fiscale                                                                      2023/2024          2023/2024          2023/2024          2023/2024          2023/2024
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                                     NaN                NaN                NaN                NaN                NaN
Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                                  2

### Record Granularity — Documentation Review

**Correction supported by documentation:** the initial interpretation that every row is an individual journal voucher item was too narrow. The published structure accommodates **journal-voucher entries and accounting/control summary records**. A row should be treated as an accounting record whose underlying detail depends on its type, not automatically as one original payment, obligation, or cash movement.

The [official XML dictionary](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/SGBT-BCMS-dd.xml) defines the four relevant fields as follows (paraphrased; XML heading IDs identify the definitions):

| Field | Documented meaning | Consequence for granularity |
|---|---|---|
| Accounting Control Number (`h9`) | Identifies an accounting transaction prepared for general-ledger posting; several transactions may be grouped under it | Repetition is allowed by the documented meaning; it is not a row identifier by itself |
| Journal Voucher Identifier (`h76`) | Optional voucher reference assigned by the Common Departmental Financial System | Absence does not prove a defective record or establish its type |
| Journal Voucher Item Identifier (`h78`) | Sequence number for an individual general-ledger entry within a voucher | Identifies voucher-entry detail, not necessarily one underlying business event |
| Accounting Summary Transaction Line Count (`h81`) | Number of detail transactions represented by a grouping; may be blank for voucher transactions | A summary can represent multiple detail transactions; this is not simply a CSV-row count |

The [dataset-specific supporting documentation](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/tcrg-rgat/soutien-support-eng.html) explicitly distinguishes journal-voucher transactions from control data in its Department Number instructions. This confirms that a uniform voucher-only interpretation is inappropriate.

**Existing local evidence:** the displayed sample already contains summary counts greater than one. That supports grouped detail in the local extract without any new analysis. The documentation does not provide a complete row-classification algorithm or a mapping from every local row to its underlying source events. Those remain unverified. The observed composite-key result is unchanged and does not establish uniform business granularity.


## 5. Data Types

This section examines the data type assigned to each attribute by Pandas.

Data types provide useful technical information, but they do not necessarily indicate the business meaning of a field.

For example, a department identifier may be stored as a numeric value while still representing a categorical dimension rather than a quantitative measure.

In [7]:
df.dtypes

Accounting-Control-Number-Numéro-contrôle-comptable                                        int64
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                                   str
Fiscal-Year-Année-Fiscale                                                                    str
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                         float64
Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                                 str
DepartmentNumber-Numéro-de-Ministère                                                     float64
Fiscal-Month-Mois-Fiscal                                                                   int64
Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal            int64
Financial-Coding-Block-Bloc-de-codage-financier                                              str
Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire                             float64
Credit/Debit-Code-Code-Crédit/

## 6. Missing Values

Missing-value analysis is used to identify attributes with incomplete coverage.

Both the absolute number and percentage of missing values are evaluated.

At this stage, columns are not removed. The purpose is only to document their completeness and determine whether they may be useful in later stages of the project.

In [8]:
missing = pd.DataFrame(
    {
        "missing_count": df.isna().sum(),
        "missing_percent": (df.isna().mean() * 100).round(2),
    }
)

missing.sort_values("missing_percent", ascending=False)

,missing_count,missing_percent
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal,9104,98.08
General-Ledger-Account-Code-Code-du-compte-du-grand-livre-général,9091,97.94
Financial-Coding-Block-Bloc-de-codage-financier,9000,96.96
Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire,2186,23.55
GBS-Control-Number-Numéro-de-contrôle-du-SBG,239,2.57
Control-Data-Type-Type-de-données-de-contrôle,178,1.92
DepartmentNumber-Numéro-de-Ministère,26,0.28
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal,26,0.28
Accounting-Summary-Transaction-Line-Count-Sommaire-comptable-transaction-compte-ligne,26,0.28
Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal,0,0.00


### Preliminary Observations

Some attributes have extremely low coverage, including fields with more than 95% missing values.

These fields may have limited analytical usefulness, but they will not be removed during the inventory stage.

Other attributes, including accounting dates, fiscal periods, debit/credit classification, and transaction amounts, appear to have significantly better coverage.

## 7. Attribute Cardinality

Cardinality represents the number of distinct values contained in each attribute.

This analysis helps identify:

- possible identifiers;
- categorical dimensions;
- low-cardinality classification fields;
- continuous or highly variable measures.

High cardinality does not automatically mean that a column is a unique identifier. Possible keys must be validated separately.

In [9]:
df.nunique().sort_values(ascending=False)

Accounting-Control-Number-Numéro-contrôle-comptable                                      9134
Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal                     8276
GBS-Control-Number-Numéro-de-contrôle-du-SBG                                             8214
Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire                              354
Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                              188
Accounting-Summary-Transaction-Line-Count-Sommaire-comptable-transaction-compte-ligne      78
DepartmentNumber-Numéro-de-Ministère                                                       54
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                                 39
Control-Data-Type-Type-de-données-de-contrôle                                              34
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                           30
Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-

### Preliminary Observations

Some attributes have a number of distinct values close to the total number of records, suggesting that they may function as identifiers or control numbers.

Other attributes contain only a few dozen distinct values, suggesting categorical dimensions such as department, journal voucher type, or control type.

Potential identifiers will be tested for uniqueness later in the notebook.

## 8. Distribution of Key Categorical Attributes

This section examines the frequency distribution of selected categorical attributes.

The objective is to understand the composition of the dataset and identify dominant categories, rare categories, and potentially useful dimensions for financial analysis.

In [10]:
df["Credit/Debit-Code-Code-Crédit/Débit"].value_counts(dropna=False)

Credit/Debit-Code-Code-Crédit/Débit
DR    7231
CR    2051
Name: count, dtype: int64

In [11]:
df["Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal"].value_counts(
    dropna=False
)

Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal
ADEPT    3439
JVEFI    2289
JVFIR     563
DOMFR     386
JVIVI     188
EPAYR     187
JVBDP     187
JVRDI     187
JVRGI     187
RDBOA     187
JVLVP     187
RGEPA     186
JVIVR     178
JVFID     176
GCMO      148
JVBRT     134
JVFRD     120
JVEFR      74
AFDEP      61
JVRFI      48
NaN        26
RDPCM      22
CGAGR      16
RGEIC      11
CADJU      11
JVBRC      10
UNSPD      10
FMSPM       9
RMSCM       9
MDDEP       9
RDSDT       7
BNKFE       6
LVTSR       5
CRCPC       4
GCRCD       4
FORFR       3
SUSPD       2
RCDSP       2
RCDCP       2
CITYE       2
Name: count, dtype: int64

In [12]:
df["Fiscal-Year-Année-Fiscale"].value_counts(dropna=False)

Fiscal-Year-Année-Fiscale
2023/2024    9225
2022/2023      57
Name: count, dtype: int64

In [13]:
df["DepartmentNumber-Numéro-de-Ministère"].value_counts(dropna=False)

DepartmentNumber-Numéro-de-Ministère
122.0    2390
97.0     2101
14.0     1436
18.0     1318
6.0       860
5.0       456
30.0       82
163.0      51
35.0       50
127.0      50
165.0      49
41.0       39
86.0       34
7.0        31
119.0      27
1.0        26
NaN        26
148.0      25
130.0      19
79.0       17
87.0       16
85.0       15
34.0       15
22.0       14
33.0       14
95.0       14
21.0       13
50.0       13
54.0        9
47.0        7
67.0        7
136.0       7
56.0        6
191.0       5
37.0        4
88.0        4
11.0        4
124.0       3
135.0       3
53.0        3
144.0       2
61.0        2
9.0         2
52.0        2
62.0        1
139.0       1
142.0       1
27.0        1
170.0       1
13.0        1
195.0       1
25.0        1
96.0        1
180.0       1
123.0       1
Name: count, dtype: int64

In [14]:
df["Control-Data-Type-Type-de-données-de-contrôle"].value_counts(dropna=False)

Control-Data-Type-Type-de-données-de-contrôle
COK    3825
CMC    2289
JMT     563
JMZ     208
CMD     188
JMR     187
JMY     187
CMI     187
JMG     187
JND     187
JML     187
NaN     178
CMF     178
JMU     176
CMT     134
JMV     120
CME      74
COL      64
JMW      48
COH      16
JNC      11
COF      11
CMJ      10
CMA      10
JOD       9
JOE       9
COJ       9
JNE       7
COP       6
JMS       5
COE       4
CMB       2
CNU       2
CNS       2
JOB       2
Name: count, dtype: int64

### Preliminary Observations

The dataset contains both debit and credit accounting entries.

Journal voucher types show several distinct accounting categories, although some types are considerably more frequent than others.

Most records are concentrated in a single fiscal year, which may limit the dataset's usefulness for long-term time-series analysis and forecasting.

## 9. Temporal Coverage

Temporal coverage is evaluated to determine the actual period represented by the dataset.

This is especially important for financial trend analysis, anomaly detection, seasonality analysis, and forecasting.

The accounting effective date is converted to a datetime representation for analysis without modifying the original raw dataset.

In [15]:
dates = pd.to_datetime(
    df["Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable"], errors="coerce"
)

In [16]:
dates.isna().sum()

np.int64(0)

In [17]:
dates.min(), dates.max()

(Timestamp('2023-04-03 00:00:00'), Timestamp('2024-01-03 00:00:00'))

In [18]:
print(df["Fiscal-Month-Mois-Fiscal"].value_counts().sort_index().to_string())

Fiscal-Month-Mois-Fiscal
1      961
2     1058
3     1021
4      793
5     1109
6     1003
7     1132
8     1079
9     1009
10      60
12      57


### Temporal Coverage Assessment

The available time range should be evaluated to determine whether the dataset provides sufficient historical depth for each Financial Manager AI capability.

A dataset may be suitable for descriptive analysis or anomaly detection while still being insufficient for reliable long-term forecasting.

### Preliminary Observations

The accounting effective dates range from April 3, 2023 to January 3, 2024.

The dataset therefore contains less than one full year of observed accounting activity.

Most observations are distributed across the available fiscal months, although the number of records is not uniform across periods.

### Fiscal Calendar — Documented Convention and Remaining Alignment Question

The dataset-linked [XML dictionary](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/SGBT-BCMS-dd.xml) defines Fiscal Month (`h17`) as the transaction's fiscal month, with permitted values 1–12; it does **not** enumerate calendar-month names. Fiscal Year (`h16`) identifies the associated fiscal year, while Accounting Effective Date (`h72`) identifies the transaction's effective date. These definitions do not prescribe how prior-year periods must relate to effective dates.

The same publisher's official [Receiver General Payment Volumes supporting documentation, “Fiscal Months”](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/paiements-payments/soutien-support-eng.html) explicitly supplies this mapping:

| Fiscal Month | Calendar month |
|---|---|
| 1 | April |
| 2 | May |
| 3 | June |
| 4 | July |
| 5 | August |
| 6 | September |
| 7 | October |
| 8 | November |
| 9 | December |
| 10 | January |
| 11 | February |
| 12 | March |

**Evidence boundary:** the table is an officially documented Receiver General convention, not a mapping inferred from the sample dates. Its application to this accounting dataset is a cross-document interpretation: the dataset's own dictionary names a fiscal month but does not repeat the table. Under that convention, fiscal year 2023/2024 spans April 2023–March 2024.

**Still unresolved:** existing outputs show 57 records labeled fiscal year 2022/2023, and separately 57 records labeled fiscal month 12, despite effective dates starting April 3, 2023. Matching counts do not establish that they are the same records. The documents reviewed do not explain this local alignment or prove late postings/adjustments. No row-level reconciliation was performed, and no dates or fiscal labels should be corrected from this documentation alone.

The relatively short historical coverage may be sufficient for descriptive financial analysis and some anomaly-detection experiments, but it may limit the reliability of long-term time-series forecasting.

## 10. Financial Measure Analysis

The main monetary attribute is analyzed to understand its statistical distribution.

The objective is to identify the scale, variability, central tendency, and extreme values of accounting transaction amounts.

Extreme values are not automatically considered errors because large financial transactions may be legitimate.

In [19]:
df["Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal"].describe()

count    9.282000e+03
mean     3.298092e+08
std      1.887057e+09
min      0.000000e+00
25%      5.782925e+02
50%      1.767720e+03
75%      6.725283e+04
max      3.955579e+10
Name: Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal, dtype: float64

### Preliminary Observations

The transaction amount distribution is highly right-skewed.

The median transaction amount is approximately 1,767.72, while the mean is substantially higher due to a relatively small number of extremely large accounting entries.

The maximum observed transaction amount is approximately 39.56 billion.

This suggests the presence of strong outliers or legitimately large accounting transactions. These observations should not be removed automatically and will require further investigation during the EDA and anomaly-detection stages.

Because of the highly skewed distribution, the median may provide a more representative measure of a typical transaction than the arithmetic mean.





### Debit and Credit Comparison

Transaction amounts are also analyzed separately by debit and credit classification.

This provides a clearer understanding of the monetary behavior of the two accounting entry types.

In [20]:
df.groupby("Credit/Debit-Code-Code-Crédit/Débit")[
    "Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal"
].agg(["count", "sum", "mean", "median", "min", "max"])

,count,sum,mean,median,min,max
Credit/Debit-Code-Code-Crédit/Débit,,,,,,
CR,2051,5.188250e+10,2.529620e+07,1723.06,0.0,6.184370e+09
DR,7231,3.009406e+12,4.161812e+08,1785.92,0.0,3.955579e+10


### Debit and Credit Observations

Debit entries are considerably more frequent than credit entries in the dataset.

Although the median transaction amounts for debit and credit entries are relatively similar, debit entries contain substantially larger extreme values.

The total debit amount is also significantly higher than the total credit amount in this dataset.

These differences should not be interpreted directly as expenses versus revenues, because debit and credit classifications represent accounting entry directions rather than business income and expense categories.

The result suggests that debit and credit transactions may need to be analyzed separately during anomaly detection and financial analysis.

## 11. Duplicates and Potential Keys

This section evaluates complete duplicate rows and investigates columns that may function as identifiers.

A column with high cardinality is not automatically unique.

In accounting datasets, a control number may identify a journal entry containing multiple individual items, which means that a combination of attributes may be required to uniquely identify each record.

In [21]:
df.duplicated().sum()

np.int64(0)

In [22]:
df["Accounting-Control-Number-Numéro-contrôle-comptable"].duplicated().sum()

np.int64(148)

### Duplicate Analysis

No complete duplicate rows were identified in the dataset.

However, the Accounting Control Number contains 148 duplicated occurrences, confirming that this attribute cannot uniquely identify individual records by itself.

This result motivates the investigation of a composite key.

### Key Investigation

If the accounting control number is repeated across multiple records, the dataset may require a composite key.

Possible combinations of control numbers, journal voucher identifiers, and journal voucher item identifiers should be investigated before defining the final database schema.

### Composite Key Investigation

The Accounting Control Number is not unique by itself.

This test evaluates whether combining the Accounting Control Number with the Journal Voucher Item Identifier produces a unique identifier for each record.

In [23]:
df.duplicated(
    subset=[
        "Accounting-Control-Number-Numéro-contrôle-comptable",
        "Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal",
    ]
).sum()

np.int64(0)

### Composite Key Result

The combination of Accounting Control Number and Journal Voucher Item Identifier produced zero duplicate records.

This indicates that the two attributes together uniquely identify each row in the current dataset.

Therefore, this combination is a strong candidate for a composite key.

The final database key should still be confirmed during the schema-design stage and validated against the official dataset documentation.

## 12. Attribute Classification

The attributes are classified according to their likely analytical role.

The main categories used are:

- **Identifier:** used to distinguish records, transactions, accounts, or control entities;
- **Dimension:** used to categorize, filter, group, or segment financial information;
- **Measure:** quantitative value that can be aggregated or statistically analyzed;
- **Control Field:** operational or accounting metadata used for validation, traceability, or internal control.

This classification is preliminary and may change after reviewing the official data dictionary.

| Attribute | Preliminary Classification | Description |
|---|---|---|
| Accounting Control Number | Identifier / Control Field | Accounting transaction reference that may group multiple transactions (dictionary `h9`) |
| Journal Voucher Type Code | Dimension | Type of accounting journal entry |
| Fiscal Year | Temporal Dimension | Fiscal year associated with the entry |
| Journal Voucher Identifier | Identifier | Optional Common Departmental Financial System voucher reference (`h76`) |
| Accounting Effective Date | Temporal Dimension | Effective accounting date |
| Department Number | Organizational Dimension | Department or organizational unit |
| Fiscal Month | Temporal Dimension | Fiscal accounting month |
| Journal Voucher Item Identifier | Identifier | Sequence of a general-ledger entry within a voucher (`h78`) |
| Financial Coding Block | Accounting Dimension | Financial classification or coding information |
| Subledger Account Identifier | Accounting Dimension | Subledger account reference |
| Credit/Debit Code | Financial Dimension | Indicates debit or credit |
| Journal Voucher Item Amount | Measure | Monetary value of the journal item |
| Control Data Type | Control Dimension | Classification used for accounting control |
| Accounting Summary Transaction Line Count | Control Measure | Detail-transaction volume within a grouping; may be blank for voucher transactions (`h81`) |
| GBS Control Number | Identifier / Control Field | System control reference |
| General Ledger Account Code | Accounting Dimension | General ledger account classification |
The four definitions marked with XML heading IDs were checked against the [official dictionary](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/SGBT-BCMS-dd.xml) on 2026-09-26. Other classifications retain their preliminary status.


## 13. Applicability to Financial Manager AI

The dataset is evaluated against the six core capabilities planned for the Financial Manager AI V1.

The objective is not to force a single dataset to support every module.

Instead, each dataset should be used only for the capabilities that are supported by its actual attributes, granularity, temporal coverage, and business meaning.

| Financial Manager Capability | Preliminary Assessment | Reason |
|---|---|---|
| Accounts Payable | Inadequate | The 16 listed attributes contain no explicit supplier, invoice, due date, payment status, or outstanding balance fields. Accounting entries alone do not establish obligations, settlement, or overdue amounts. |
| Financial Data Analysis | Suitable | The 9,282 records include complete accounting effective dates, fiscal periods, debit/credit codes, and item amounts. There are 54 observed departments, 39 voucher types, and 354 subledger accounts for segmentation, subject to missing values and confirmation of code meanings. This supports descriptive accounting analysis; debit/credit direction does not establish revenue or expense categories. |
| Budget Variance | Inadequate | No budget, planned amount, or budget version appears in the listed attributes. Item amounts may eventually contribute an actuals component, but budget-to-actual comparison requires a separate budget source and validated accounting mappings. |
| Financial Document RAG | Inadequate | The notebook inventories a structured CSV of accounting records, with no document text or corpus of contracts, reports, or policies. These records alone do not provide the documentary evidence required for this capability. |
| Financial Anomaly Detection | Potentially Suitable | Amounts vary strongly (median approximately 1,767.72; maximum approximately 39.56 billion), and debit/credit, department, voucher type, and account dimensions may support contextual comparisons. Rare categories, uneven category frequencies, missing account information, and limited history require investigation. No anomaly labels or validation results are present, and extreme values are not established errors. |
| Cash Flow Forecast | Partial | Dates and monetary amounts provide a possible starting point for investigating time-based accounting activity, but accounting dates and debit/credit entries are not confirmed cash movement dates and directions. No explicit cash balance, payment/receipt schedule, or cash-flow classification is listed. The observed period is less than one year, so neither a usable cash series nor reliable forecasting is established. |


### Preliminary Dataset Role

The dataset is most useful as a preliminary source for **Financial Data Analysis**, especially descriptive accounting exploration by date, fiscal period, department, voucher type, and debit/credit direction. Account-based analysis is more constrained: subledger identifiers are 23.55% missing, and general ledger account codes are 97.94% missing.

It is also a candidate for **Financial Anomaly Detection** experiments after the accounting semantics, record granularity, comparable groups, and data quality have been investigated. Large amounts and uncommon categories are investigation signals, not confirmed anomalies.

Its role in **Cash Flow Forecast** is conditional on identifying genuine cash movements and obtaining sufficient history and supporting cash information. Separate sources are needed for **Accounts Payable**, **Budget Variance**, and **Financial Document RAG**. Suitability here describes analytical potential in the inventoried file; it does not establish operational readiness for the V1.


## 14. Preliminary Conclusions

### Dataset Representation and Granularity

The Receiver General Accounting Transactions file contains **9,282 records and 16 attributes** describing accounting journal activity. The documentation review in section 4 corrects the uniform journal-voucher-item assumption: the source structure accommodates voucher entries and accounting/control summary records. The existing sample includes summary counts greater than one. Exact local row types remain unclassified; one row is not necessarily one underlying business transaction, obligation, payment, or cash movement.

### Useful Dimensions and Measures

Useful dimensions include accounting effective date (188 distinct dates), fiscal year and fiscal month, department (54 distinct values), journal voucher type (39), control data type (34), debit/credit direction (2), and subledger account (354). Financial coding blocks and general ledger account codes are additional possible accounting dimensions, but have very limited coverage. Counts of distinct values do not establish the meanings of the codes.

Journal Voucher Item Amount is the main monetary measure. Accounting Summary Transaction Line Count measures grouped detail-transaction volume according to the official dictionary (section 4); whether totals across local record types would overlap remains unverified. Amounts range from zero to approximately 39.56 billion, with a median of approximately 1,767.72 and a mean of approximately 329.81 million. The large mean–median difference and extreme maximum indicate pronounced skew. The notebook does not establish currency or monetary units, and debit/credit direction must not be interpreted directly as revenue/expense or cash inflow/outflow.

### Data Quality and Missingness

The existing checks found **no complete duplicate rows** and **no failed accounting-date conversions**. Accounting control number, item identifier, accounting effective date, fiscal year, fiscal month, debit/credit code, and item amount have no missing values. These observations establish completeness and parseability for those checks, not full business validity.

Missingness substantially limits several fields:

- Journal Voucher Identifier: **98.08%** missing; General Ledger Account Code: **97.94%**; Financial Coding Block: **96.96%**.
- Subledger Account Identifier: **23.55%** missing, limiting account-based comparisons.
- GBS Control Number: **2.57%** missing; Control Data Type: **1.92%**.
- Department Number, Journal Voucher Type Code, and Accounting Summary Transaction Line Count: **0.28%** missing each.

Several identifiers are stored as floating-point values, while accounting effective dates are stored as strings in the loaded data. Appropriate representations and identifier fidelity need later assessment. Missing fields should not automatically be discarded or imputed. The documentation permits some type-dependent blanks, including summary counts for voucher transactions (section 4); it does not establish whether every local absence follows those rules.

### Temporal Coverage and Candidate Key

Accounting effective dates span **April 3, 2023 to January 3, 2024**, less than one full year. Fiscal-year labels contain 9,225 records for 2023/2024 and 57 for 2022/2023. Fiscal-month values cover 1–10 and 12, with only 60 records in month 10, 57 in month 12, and none shown for month 11. Section 9 records the officially documented Receiver General convention of fiscal months 1–12 corresponding to April–March, with its cross-document evidence boundary. Local fiscal labels still require reconciliation with effective dates; the documents do not establish whether the pattern reflects adjustments, extract boundaries, or missing coverage. Continuous coverage and repeatable annual seasonality have not been demonstrated.

Accounting Control Number alone has **148 duplicated occurrences**. **Accounting Control Number + Journal Voucher Item Identifier** has zero duplicate combinations, and both fields are complete. This is a candidate composite key for the current file, subject to confirmation of business meaning and stability across future extracts or periods.

### Capability Implications

The dataset supports descriptive **Financial Data Analysis** and may support exploratory **Financial Anomaly Detection**. Its skewed amounts, debit/credit differences, and uneven department and voucher-type frequencies motivate comparisons within appropriate accounting groups. Extreme or rare observations may be legitimate, and missingness or differences in aggregation may create misleading signals. No anomaly labels or evidence of detection performance are available in the notebook.

**Accounts Payable** lacks explicit obligations, suppliers, invoices, due dates, and settlement status. **Budget Variance** lacks budget amounts and versions, even if journal amounts could later supply appropriately mapped actuals. **Financial Document RAG** lacks a documentary corpus. **Cash Flow Forecast** lacks confirmed cash movements, opening balances, and scheduled receipts/payments; the short history further limits temporal evaluation. This file alone does not establish support for those complete capabilities.

### Later Investigation During Detailed EDA and Cleaning

Once separately authorized, later work should investigate:

- Apply the now-linked official dictionary to local record types; investigate remaining code meanings, currency/units, sign conventions, and possible overlap between voucher entries and grouped detail. The dictionary establishes summary-count meaning, not a complete local grouping model.
- Fiscal-year/month alignment with effective dates, coverage gaps, extract boundaries, and whether additional history is available.
- Missingness by voucher type, department, period, and control type; whether absent fields are structurally inapplicable or incomplete; and justified treatment of each field.
- Identifier representation, relationships among control numbers and voucher items, and composite-key stability beyond this file.
- Zero and extreme amounts, debit/credit differences, and appropriate comparison groups before treating any observation as anomalous or erroneous.
- Whether any entries can be reliably mapped to revenue, expense, obligations, payments, or cash movements without double counting, and which additional sources are needed for the unsupported capabilities.

These are investigation topics, not cleaning decisions or implementation approval. This documentation adds no analyses, transformations, cleaning, modeling, or raw-data changes.


# Company Financials Dataset

This second inventory profiles the supplied file independently of the completed Receiver General inventory. All original headers and values are retained. Profiling summaries do not change the raw file or the loaded records; no cleaning, feature engineering, modeling, or operational implementation is performed.

## 1. Imports

Reuse `pandas as pd` from the first inventory. The following cells use a separate `company_financials` DataFrame so the Receiver General variables and analyses remain intact. The standard-library `re` module validates formatted amounts for explicitly authorized temporary numeric profiling; no new dependencies are required.

In [24]:
import re

## 2. Dataset Loading

- **Dataset label:** Company Financials Dataset.
- **Original file:** `Financials.csv`.
- **Format:** comma-separated CSV with a header row and quoted fields containing commas.
- **Raw-data path (from this notebook):** `../data/raw/Company Financials Dataset/Financials.csv`.
- **Local file selection:** one file was found in this dataset directory; no competing primary file was present.
- **Source/origin:** not documented in the inspected local project documentation or supplied CSV. No external lookup was performed.
- **Local documentation/data dictionary:** none found for this dataset. Column meanings below are preliminary interpretations of headers and values.
- **Preservation:** the raw file is read only; its original name, headers, spacing, and values remain unchanged. Raw-file hashes are checked before and after this inventory.

The dataset label does not establish that this is a company financial-statement panel. The actual contents determine its analytical role.

In [25]:
company_financials = pd.read_csv(
    "../data/raw/Company Financials Dataset/Financials.csv"
)

## 3. General Dataset Structure

Inspect shape, exact column names, inferred types, non-null counts, and memory usage. Whitespace in headers is retained and is visible in the column-name list.

In [26]:
print("Shape:", company_financials.shape)
print("Columns:", company_financials.columns.tolist())
company_financials.info()

Shape: (700, 16)
Columns: ['Segment', 'Country', ' Product ', ' Discount Band ', ' Units Sold ', ' Manufacturing Price ', ' Sale Price ', ' Gross Sales ', ' Discounts ', '  Sales ', ' COGS ', ' Profit ', 'Date', 'Month Number', ' Month Name ', 'Year']
<class 'pandas.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Segment                700 non-null    str  
 1   Country                700 non-null    str  
 2    Product               700 non-null    str  
 3    Discount Band         700 non-null    str  
 4    Units Sold            700 non-null    str  
 5    Manufacturing Price   700 non-null    str  
 6    Sale Price            700 non-null    str  
 7    Gross Sales           700 non-null    str  
 8    Discounts             700 non-null    str  
 9     Sales                700 non-null    str  
 10   COGS                  700 non-null    str  
 11 

The file contains **700 records and 16 attributes**. Pandas infers **14 string columns and 2 integer columns** (`Month Number` and `Year`); all columns have 700 non-null entries. Reported memory usage is approximately **87.6 KB** with the default `info()` estimate, which is not a deep measurement of string storage. Most apparent financial measures are stored as text, not numeric values.

## 4. Data Sample and Record Granularity

Inspect the first and last three rows, transposed for readability. These positional samples illustrate the structure; they are not a random sample or proof of representativeness.

In [27]:
print("First three records:")
print(company_financials.head(3).T.to_string())
print("Last three records:")
print(company_financials.tail(3).T.to_string())

First three records:
                                  0             1             2
Segment                  Government    Government     Midmarket
Country                      Canada       Germany        France
 Product                 Carretera     Carretera     Carretera 
 Discount Band                None          None          None 
 Units Sold              $1,618.50     $1,321.00     $2,178.00 
 Manufacturing Price         $3.00         $3.00         $3.00 
 Sale Price                 $20.00        $20.00        $15.00 
 Gross Sales            $32,370.00    $26,420.00    $32,670.00 
 Discounts                    $-            $-            $-   
  Sales                 $32,370.00    $26,420.00    $32,670.00 
 COGS                   $16,185.00    $13,210.00    $21,780.00 
 Profit                 $16,185.00    $13,210.00    $10,890.00 
Date                     01/01/2014    01/01/2014    01/06/2014
Month Number                      1             1             6
 Month Name        

A row appears to describe a **sales observation for a segment, country, product, discount band, and monthly period**, with units, prices, sales, costs, and profit. There is no company name/ID, invoice ID, or transaction ID. This is **not established as company-year or company-quarter data**.

The row's precise aggregation level remains unknown: records may represent underlying transactions or summarized activity, and multiple rows can share the same listed dimensions and date (tested in section 11). Country and segment are not company identifiers. The date field appears to label a month rather than establish an individual sale or payment date.

## 5. Data Types

Technical storage types are inspected without converting them. Integer time labels are temporal dimensions rather than financial measures. The eight apparent financial/quantity measures remain strings because their supplied values include formatting.

In [28]:
print(company_financials.dtypes.to_string())
print("Headers with surrounding whitespace:", sum(c != c.strip() for c in company_financials.columns))

Segment                    str
Country                    str
 Product                   str
 Discount Band             str
 Units Sold                str
 Manufacturing Price       str
 Sale Price                str
 Gross Sales               str
 Discounts                 str
  Sales                    str
 COGS                      str
 Profit                    str
Date                       str
Month Number             int64
 Month Name                str
Year                     int64
Headers with surrounding whitespace: 11


**11 headers** contain surrounding whitespace. This check detects formatting; it does not rename columns. Product, discount-band, and month-name samples also contain surrounding whitespace.

`Units Sold` appears semantically quantitative, while Manufacturing Price, Sale Price, Gross Sales, Discounts, Sales, COGS, and Profit appear to be financial measures. Their text includes dollar signs, comma grouping, and sometimes parentheses or dash placeholders. Even Units Sold has a dollar sign in the supplied file, so its meaning and formatting need confirmation. A `$` symbol does not identify the currency, and BRL compatibility is not established. `Date` remains text, and the original DataFrame retains every original type and value. Section 10 uses a separate temporary numeric view solely for the profiling explicitly authorized by the user.

## 6. Missing Values

Count missing values as recognized by Pandas and separately check whitespace-only strings and exact dollar-dash formatting. A formatted placeholder can escape the usual null count; its meaning is not assumed.

In [29]:
company_missing = pd.DataFrame({
    "missing_count": company_financials.isna().sum(),
    "missing_percent": (company_financials.isna().mean() * 100).round(2),
})
print(company_missing.to_string())
print("Whitespace-only string counts:")
print(pd.Series({c: company_financials[c].str.fullmatch(r"\s*", na=False).sum()
                 for c in company_financials.select_dtypes(include=["str", "object"]).columns}).to_string())
print("Dollar-dash placeholders:")
for company_column in [" Discounts ", " Profit "]:
    company_dash = company_financials[company_column].str.fullmatch(r"\s*\$-\s*", na=False)
    print(repr(company_column), int(company_dash.sum()), f"{company_dash.mean() * 100:.2f}%")

                       missing_count  missing_percent
Segment                            0              0.0
Country                            0              0.0
 Product                           0              0.0
 Discount Band                     0              0.0
 Units Sold                        0              0.0
 Manufacturing Price               0              0.0
 Sale Price                        0              0.0
 Gross Sales                       0              0.0
 Discounts                         0              0.0
  Sales                            0              0.0
 COGS                              0              0.0
 Profit                            0              0.0
Date                               0              0.0
Month Number                       0              0.0
 Month Name                        0              0.0
Year                               0              0.0
Whitespace-only string counts:
Segment                  0
Country                 

All 16 fields have **0 missing values (0.00%)** under Pandas' default CSV parsing, and no whitespace-only strings were found. There are no partially populated or extremely sparse fields by that measure.

However, Discounts contains **53 dollar-dash placeholders (7.57%)** and Profit contains **5 (0.71%)**. They are non-null strings. They may represent formatted zeros or another convention; this inventory does not replace them with zeros or missing values. The `None` discount-band label is an observed category, not a demonstrated missing value. Structural missingness cannot be inferred from null counts alone.

## 7. Attribute Cardinality

Distinct counts exclude nulls by default. Counts for formatted measures describe distinct strings, not validated numeric values. Cardinality alone does not establish keys.

In [30]:
print(company_financials.nunique().sort_values(ascending=False).to_string())

  Sales                  559
 Profit                  557
 Gross Sales             550
 COGS                    545
 Discounts               515
 Units Sold              510
Date                      16
Month Number              12
 Month Name               12
 Sale Price                7
 Product                   6
 Manufacturing Price       6
Segment                    5
Country                    5
 Discount Band             4
Year                       2


The dataset has **5 segments, 5 countries, 6 products, 4 discount bands, 16 date strings, 12 month numbers/names, and 2 years**. Financial text fields range from 6 distinct Manufacturing Price strings to 559 Sales strings; Units Sold has 510. No column has 700 distinct values. Temporal and sales dimensions are potential composite-key components, subject to direct duplicate testing.

## 8. Distribution of Key Categorical Attributes

Inspect segment, country, product, and discount band. High-cardinality measure strings are not treated as categorical business identifiers.

In [31]:
for company_column in ["Segment", "Country", " Product ", " Discount Band "]:
    print(repr(company_column))
    print(company_financials[company_column].value_counts(dropna=False).to_string())

'Segment'
Segment
Government          300
Midmarket           100
Channel Partners    100
Enterprise          100
Small Business      100
'Country'
Country
Canada                      140
Germany                     140
France                      140
Mexico                      140
United States of America    140
' Product '
 Product 
Paseo         202
Velo          109
VTT           109
Amarilla       94
Carretera      93
Montana        93
' Discount Band '
 Discount Band 
High       245
Medium     242
Low        160
None        53


Government is the largest segment (**300 rows**); each of the four other segments has **100**. Each country has **140 rows**, so country counts are balanced in this file. Product counts are uneven: Paseo has **202**, Velo and VTT **109 each**, Amarilla **94**, and Carretera and Montana **93 each**.

Discount bands are High **245**, Medium **242**, Low **160**, and None **53**. None is the least frequent band, but rarity alone is not evidence of an error or anomaly. These are record counts, not shares of revenue, customers, or companies. The exact business definitions of the segments and discount bands remain undocumented.

## 9. Temporal Coverage

Inspect existing Year and Month Number labels without type conversion. Compare Date strings to an expected first-of-month `DD/MM/YYYY` representation and Month Name to the supplied month number. These are consistency checks, not rewritten date fields.

In [32]:
print("Year counts:")
print(company_financials["Year"].value_counts().sort_index().to_string())
print("Year-by-month record counts:")
print(pd.crosstab(company_financials["Year"], company_financials["Month Number"]).to_string())
company_periods = sorted(set(zip(company_financials["Year"], company_financials["Month Number"])))
print("Observed (year, month) periods:", company_periods)
print("Period count:", len(company_periods))
print("Date string count:", company_financials["Date"].nunique())
print("Date strings aligned with first-of-month DD/MM/YYYY:", sum(
    date == f"01/{month:02d}/{year}"
    for date, month, year in zip(company_financials["Date"], company_financials["Month Number"], company_financials["Year"])
))
company_month_names = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
print("Month names matching supplied month numbers (including original spaces):", sum(
    name == f" {company_month_names[month - 1]} "
    for name, month in zip(company_financials[" Month Name "], company_financials["Month Number"])
))
for company_column in ["Country", "Segment", " Product "]:
    print("Distinct months per year by", repr(company_column))
    print(company_financials.groupby([company_column, "Year"])["Month Number"].nunique().unstack().to_string())

Year counts:
Year
2013    175
2014    525
Year-by-month record counts:
Month Number  1   2   3   4   5   6   7   8   9   10  11  12
Year                                                        
2013           0   0   0   0   0   0   0   0  35  70  35  35
2014          35  35  35  35  35  70  35  35  35  70  35  70
Observed (year, month) periods: [(2013, 9), (2013, 10), (2013, 11), (2013, 12), (2014, 1), (2014, 2), (2014, 3), (2014, 4), (2014, 5), (2014, 6), (2014, 7), (2014, 8), (2014, 9), (2014, 10), (2014, 11), (2014, 12)]
Period count: 16
Date string count: 16
Date strings aligned with first-of-month DD/MM/YYYY: 700
Month names matching supplied month numbers (including original spaces): 700
Distinct months per year by 'Country'
Year                      2013  2014
Country                             
Canada                       4    12
France                       4    12
Germany                      4    12
Mexico                       4    12
United States of America     4    12


Observed periods run from **September 2013 through December 2014**, comprising **16 consecutive monthly labels** with no month absent inside that observed interval. There are **175 rows in 2013** (September–December) and **525 in 2014** (January–December). Earlier 2013 months lie outside the observed range; their absence is not established as a data defect.

All **700** Date strings match `01/MM/YYYY` using their Month Number and Year, and all **700** month names match the month labels. Together these support interpreting the endpoints as **September 1, 2013 and December 1, 2014**, with dates serving as monthly labels. They do not prove exact sale dates or fiscal-year boundaries. There is no explicit fiscal-year field, so fiscal/calendar equivalence is not assumed.

Period counts are uneven: most months have **35 rows**, while October 2013 and June, October, and December 2014 have **70 each**. Every observed country, segment, and product is represented in all 16 periods individually; this does not prove that every cross-combination is present or that volumes are balanced. Coverage across companies cannot be assessed because there is no company identifier. This is a short monthly sales history, not an identified company-year/company-quarter panel.

## 10. Financial Measure Analysis

The apparent measures are Units Sold, Manufacturing Price, Sale Price, Gross Sales, Discounts, Sales, COGS, and Profit. They are all stored as strings. First, `describe()` reports **text counts, distinct counts, most frequent text, and frequency**. A format profile inspects dollar signs, comma grouping, parentheses, and placeholders. Then an explicitly authorized temporary numeric view supports descriptive amount statistics without modifying the original DataFrame or raw file.

In [33]:
company_measures = [" Units Sold ", " Manufacturing Price ", " Sale Price ", " Gross Sales ", " Discounts ", "  Sales ", " COGS ", " Profit "]
print("Text descriptive statistics (not numeric amount statistics):")
print(company_financials[company_measures].describe().T.to_string())
print("Original financial-value format profile:")
print(pd.DataFrame({
    "dollar_sign_count": {c: company_financials[c].str.contains("$", regex=False).sum() for c in company_measures},
    "two_or_more_commas": {c: (company_financials[c].str.count(",") >= 2).sum() for c in company_measures},
    "parentheses_count": {c: company_financials[c].str.contains("(", regex=False).sum() for c in company_measures},
    "dollar_dash_count": {c: company_financials[c].str.fullmatch(r"\s*\$-\s*", na=False).sum() for c in company_measures},
}).to_string())
print("Examples with multiple comma separators (not numerically ranked):")
print(company_financials.loc[company_financials[" Gross Sales "].str.count(",") >= 2, [" Gross Sales ", "  Sales "]].head().to_string(index=False))
print("Parenthesized Profit examples (not numerically ranked):")
print(company_financials.loc[company_financials[" Profit "].str.contains("(", regex=False), [" Profit "]].head().to_string(index=False))

Text descriptive statistics (not numeric amount statistics):
                     count unique           top freq
Units Sold             700    510      $727.00     5
Manufacturing Price    700      6       $10.00   202
Sale Price             700      7       $20.00   100
Gross Sales            700    550   $37,050.00     3
Discounts              700    515         $-      53
 Sales                 700    559   $32,670.00     2
COGS                   700    545   $17,430.00     4
Profit                 700    557         $-       5
Original financial-value format profile:


                      dollar_sign_count  two_or_more_commas  parentheses_count  dollar_dash_count
Units Sold                          700                   0                  0                  0
Manufacturing Price                 700                   0                  0                  0
Sale Price                          700                   0                  0                  0
Gross Sales                         700                 268                  0                  0
Discounts                           700                  15                  0                 53
 Sales                              700                 263                  0                  0
COGS                                700                 258                  0                  0
Profit                              700                  52                 58                  5
Examples with multiple comma separators (not numerically ranked):
  Gross Sales          Sales 
 $5,29,550.00   $5,29,

### Temporary Numeric Profiling — Explicitly Authorized

The user authorized temporary numeric parsing for profiling only. The original DataFrame and CSV remain unchanged. Parsing uses these explicit conventions:

- A decimal point separates two fractional digits. Commas in validated three-digit or Indian-style two/three-digit groups are interpreted as grouping separators; surrounding spaces and `$` are formatting, not currency identification.
- Matched parentheses are interpreted as negative amounts for this profile. This is an accounting-format assumption, not evidence of an error or verified source semantics.
- Dollar-dash strings remain **unresolved**. They are represented as unavailable only in the temporary numeric view and excluded from descriptive numeric statistics; they are not imputed as zero and original strings remain intact.
- Unexpected formats cause an error instead of silently discarding values. Units Sold is parsed as a magnitude for exploration only; its dollar formatting and fractional units still need business clarification.

Statistics therefore cover 700 numeric values for most measures, **647 for Discounts**, and **695 for Profit**. Floating-point arithmetic here is descriptive profiling, not an approved operational monetary representation.

In [34]:
company_numeric_profile = pd.DataFrame(index=company_financials.index)
company_amount_pattern = (
    r"\s*\$\s*(?P<open>\()?\s*"
    r"(?P<amount>(?:\d{1,3}(?:,\d{3})+|\d{1,2}(?:,\d{2})*,\d{3}|\d+)\.\d{2})"
    r"\s*(?P<close>\))?\s*"
)
for company_column in company_measures:
    company_numeric_values = []
    for company_raw_value in company_financials[company_column]:
        if re.fullmatch(r"\s*\$-\s*", company_raw_value):
            company_numeric_values.append(float("nan"))
            continue
        company_match = re.fullmatch(company_amount_pattern, company_raw_value)
        if company_match is None or bool(company_match["open"]) != bool(company_match["close"]):
            raise ValueError(f"Unrecognized profiling format: {company_raw_value!r}")
        company_magnitude = float(company_match["amount"].replace(",", ""))
        company_numeric_values.append(-company_magnitude if company_match["open"] else company_magnitude)
    company_numeric_profile[company_column] = company_numeric_values
print("Numeric descriptive statistics; dollar-dash placeholders excluded:")
print(company_numeric_profile.describe().T.to_string(float_format=lambda value: f"{value:,.2f}"))
print("Distribution profile under the stated parsing conventions:")
print(pd.DataFrame({
    "skewness": company_numeric_profile.skew(),
    "negative_count": (company_numeric_profile < 0).sum(),
    "explicit_numeric_zero_count": (company_numeric_profile == 0).sum(),
    "unresolved_placeholder_count": company_numeric_profile.isna().sum(),
}).to_string(float_format=lambda value: f"{value:,.3f}"))

Numeric descriptive statistics; dollar-dash placeholders excluded:
                      count       mean        std        min       25%       50%        75%          max
Units Sold           700.00   1,608.29     867.43     200.00    905.00  1,542.50   2,229.12     4,492.50
Manufacturing Price  700.00      96.48     108.60       3.00      5.00     10.00     250.00       260.00
Sale Price           700.00     118.43     136.78       7.00     12.00     20.00     300.00       350.00
Gross Sales          700.00 182,759.43 254,262.28   1,799.00 17,391.75 37,980.00 279,025.00 1,207,500.00
Discounts            647.00  14,227.59  23,562.83      18.41  1,061.00  3,108.00  18,809.25   149,677.50
 Sales               700.00 169,609.07 236,726.35   1,655.08 15,928.00 35,540.20 261,077.50 1,159,200.00
COGS                 700.00 145,475.21 203,865.51     918.00  7,490.00 22,506.25 245,607.50   950,625.00
Profit               695.00  24,307.49  42,865.12 -40,617.50  2,929.32  9,297.00  22,777.50  

In [35]:
print("Observed minima and maxima with original text and context; not error labels:")
for company_column in [" Gross Sales ", " Discounts ", "  Sales ", " COGS ", " Profit "]:
    print(repr(company_column))
    company_extreme_indices = [company_numeric_profile[company_column].idxmin(), company_numeric_profile[company_column].idxmax()]
    print(company_financials.loc[company_extreme_indices, ["Segment", "Country", " Product ", " Discount Band ", "Date", company_column]].to_string())

Observed minima and maxima with original text and context; not error labels:
' Gross Sales '
        Segment                   Country  Product   Discount Band         Date     Gross Sales 
359  Government                    Canada    Paseo          Medium   01/05/2014       $1,799.00 
192  Government  United States of America    Paseo             Low   01/07/2014   $12,07,500.00 
' Discounts '
        Segment  Country  Product   Discount Band         Date      Discounts 
88   Government  Germany      VTT             Low   01/03/2014         $18.41 
673  Government   Mexico    Paseo            High   01/05/2014   $1,49,677.50 
'  Sales '
        Segment                   Country  Product   Discount Band         Date           Sales 
359  Government                    Canada    Paseo          Medium   01/05/2014       $1,655.08 
192  Government  United States of America    Paseo             Low   01/07/2014   $11,59,200.00 
' COGS '
              Segment Country    Product   Discount Ba

### Numerical Distribution Observations

Values below are in **unconfirmed source units**, not asserted BRL or another specific currency. Discounts and Profit summaries exclude their unresolved placeholders.

| Measure | Numeric count | Mean | Median | Minimum | Maximum |
|---|---:|---:|---:|---:|---:|
| Units Sold | 700 | 1,608.29 | 1,542.50 | 200.00 | 4,492.50 |
| Manufacturing Price | 700 | 96.48 | 10.00 | 3.00 | 260.00 |
| Sale Price | 700 | 118.43 | 20.00 | 7.00 | 350.00 |
| Gross Sales | 700 | 182,759.43 | 37,980.00 | 1,799.00 | 1,207,500.00 |
| Discounts | 647 | 14,227.59 | 3,108.00 | 18.41 | 149,677.50 |
| Sales | 700 | 169,609.07 | 35,540.20 | 1,655.08 | 1,159,200.00 |
| COGS | 700 | 145,475.21 | 22,506.25 | 918.00 | 950,625.00 |
| Profit | 695 | 24,307.49 | 9,297.00 | -40,617.50 | 262,200.00 |

Gross Sales, Sales, COGS, Discounts, and Profit have means substantially above medians, consistent with right-skewed distributions. Their sample skewness is approximately **1.674, 1.696, 1.549, 2.566, and 2.701**, respectively. Units Sold has a closer mean and median (skewness approximately **0.436**). The price fields have only 6 and 7 distinct supplied values, so their distributions reflect discrete price levels and category mix rather than a homogeneous continuous population.

Under the stated sign convention, **58 Profit observations are negative**. No explicitly numeric zero is found in any measure; this does not establish absence of zero activity because dollar-dash values remain unresolved. Gross Sales reaches **1,207,500.00**, Sales **1,159,200.00**, and Profit **262,200.00** in a Government / United States of America / Paseo record labeled July 2014. The minimum Profit is **-40,617.50** for Enterprise / Canada / VTT in November 2013. These are observed extremes, not confirmed errors or anomalies, and none are removed.

The file does not identify companies or their sizes. Comparisons across products, segments, or countries need to consider sales volume, price, discount, period, and aggregation differences. If company-level data is later obtained, raw amounts across companies of different sizes will likewise need contextual comparison. No normalization or financial reconciliation is performed here.

## 11. Duplicates and Potential Keys

Check full-row duplicates and combinations of existing dimensions. Duplicate counts below count occurrences after the first, rather than all rows in repeated groups. No rows or columns are removed.

In [36]:
company_dimension_key = ["Segment", "Country", " Product ", " Discount Band ", "Date"]
company_price_key = company_dimension_key + [" Sale Price "]
print("Complete duplicate rows:", company_financials.duplicated().sum())
print("Individually unique columns:", [c for c in company_financials.columns if company_financials[c].is_unique])
for company_key in [company_dimension_key, company_price_key]:
    print("Candidate:", company_key)
    print("Repeated occurrences:", company_financials.duplicated(subset=company_key).sum())
    print("Rows with missing key components:", company_financials[company_key].isna().any(axis=1).sum())
print("Examples sharing the five-dimensional combination:")
print(company_financials.loc[company_financials.duplicated(subset=company_dimension_key, keep=False), company_price_key + [" Units Sold "]].head(6).to_string(index=False))

Complete duplicate rows: 0
Individually unique columns: []
Candidate: ['Segment', 'Country', ' Product ', ' Discount Band ', 'Date']
Repeated occurrences: 12
Rows with missing key components: 0
Candidate: ['Segment', 'Country', ' Product ', ' Discount Band ', 'Date', ' Sale Price ']
Repeated occurrences: 0
Rows with missing key components: 0
Examples sharing the five-dimensional combination:
   Segment                  Country  Product   Discount Band        Date  Sale Price   Units Sold 
Government                   Canada    Paseo          Medium  01/10/2013     $350.00    $1,228.00 
Government                   Canada    Paseo          Medium  01/10/2013      $20.00    $1,389.00 
Government United States of America     Velo          Medium  01/06/2014     $350.00      $602.00 
Government United States of America      VTT          Medium  01/12/2014      $20.00    $2,663.00 
Government United States of America      VTT          Medium  01/12/2014       $7.00      $570.00 
Government 

There are **0 complete duplicate rows**, and **no individually unique column**. Segment + Country + Product + Discount Band + Date has **12 repeated occurrences**. Adding **Sale Price** results in **0 repeated combinations** and no missing key components in the current file.

This six-field combination is an **empirical candidate only**, not an approved business or database key. Sale Price is a measure, not a stable record identifier; the candidate also depends on literal text formatting. Its uniqueness may not survive additional records or later representation decisions. A source-level record identifier and the actual aggregation grain should be investigated. Repeated dimensional combinations with different prices are not automatically duplicate business events.

## 12. Attribute Classification

Classifications are preliminary because no official local data dictionary was found. Labels below omit surrounding header whitespace for readability; the actual DataFrame retains it. There is no explicit company or record identifier.

| Attribute | Preliminary Classification | Description |
|---|---|---|
| Segment | Commercial Dimension | Sales/customer segment label; not a company identifier |
| Country | Geographic Dimension | Country associated with a sales observation; precise relationship undocumented |
| Product | Product Dimension | Product label |
| Discount Band | Financial Dimension | Categorical discount level; thresholds undocumented |
| Units Sold | Quantity Measure (stored as text) | Apparent sales quantity; dollar formatting needs clarification |
| Manufacturing Price | Financial Measure (stored as text) | Apparent manufacturing price; per-unit basis and units unconfirmed |
| Sale Price | Financial Measure / Candidate-Key Component | Apparent sale price; contributes to observed uniqueness but is not an identifier by meaning |
| Gross Sales | Financial Measure (stored as text) | Apparent gross sales amount |
| Discounts | Financial Measure (stored as text) | Apparent discount amount; includes dollar-dash placeholders |
| Sales | Financial Measure (stored as text) | Apparent sales amount; gross/net relationship needs verification |
| COGS | Financial Measure (stored as text) | Label conventionally denotes cost of goods sold; definition unconfirmed |
| Profit | Financial Measure (stored as text) | Apparent profit amount; includes parentheses and dollar-dash placeholders |
| Date | Temporal Dimension (stored as text) | First-of-month label consistent with supplied month/year |
| Month Number | Temporal Dimension | Integer month label, 1–12 |
| Month Name | Temporal Dimension | Text month label consistent with Month Number |
| Year | Temporal Dimension | Year label, 2013 or 2014; no explicit fiscal definition |

## 13. Applicability to Financial Manager AI

Assessments describe the supplied data's analytical potential, not production readiness. Numerical profiling uses the explicit temporary parsing conventions in section 10; source units and business definitions remain unconfirmed.

| Financial Manager Capability | Preliminary Assessment | Reason |
|---|---|---|
| Accounts Payable | Inadequate | No supplier, purchase invoice, obligation amount, issue/due date, settlement status, outstanding balance, or linked payment records. COGS is not an obligation or payment ledger. |
| Financial Data Analysis | Suitable | Sales, discounts, costs, profit, quantity, and prices are available by product, segment, country, and 16 monthly periods. Numerical use remains conditional on validating text representations, currency, and measure definitions. There is no company identity or department field. |
| Budget Variance | Inadequate | No planned/budget amount, budget version, or budget scenario. Sales/cost/profit observations could contribute actuals only after their bases are validated and a compatible budget is supplied. |
| Financial Document RAG | Inadequate | A structured CSV supplies no contracts, policies, report text, document metadata, or versioned documentary evidence. |
| Financial Anomaly Detection | Potentially Suitable | 700 sales observations have multiple related measures and categorical/temporal context. Right-skewed sales/cost/profit distributions and 58 negative Profit values under the stated parsing convention provide investigation candidates. Comparable groups, accounting relationships, and anomaly evaluation remain unresolved; no anomaly labels are supplied. |
| Cash Flow Forecast | Partial | Sixteen consecutive monthly sales periods provide possible explanatory context, but sales and profit are not confirmed cash receipts. No actual cash movements, cash-account balances, settlement dates, or future receivable/payable schedules are listed. Short history and unconfirmed company scope further limit forecasting. |

### Preliminary Dataset Role

The strongest role is **Financial Data Analysis** of sales, costs, discounts, and profitability, with **Financial Anomaly Detection** as an exploratory candidate after representation and business-meaning checks. This is complementary to Receiver General accounting-entry analysis, not a replacement for obligations, budgets, cash records, or documents. Its potential forecasting role is contextual only until cash linkage and entity scope are established.

## 14. Preliminary Conclusions

The supplied Company Financials file appears to contain **700 sales observations with 16 attributes**, rather than an identified company financial-statement panel. The preliminary grain is a segment/country/product/discount-band/month observation; repeated combinations show that those dimensions alone do not uniquely specify a record. No company ID, source transaction ID, or invoice reference is supplied.

Important dimensions are 5 segments, 5 countries, 6 products, 4 discount bands, and 16 monthly labels. Eight apparent measures cover quantity, prices, gross sales, discounts, sales, COGS, and profit. They all remain formatted strings. Technical completeness is high: all fields have zero Pandas-recognized missing values and no whitespace-only strings, but 53 Discounts values and 5 Profit values are dollar-dash placeholders whose meaning is unresolved. Profit has 58 parenthesized values. Header/value whitespace, comma grouping, currency, signs, and the dollar-formatted quantity require later investigation.

Coverage is **September 2013–December 2014**, with no missing monthly label within that interval. All countries, segments, and products individually occur in all periods, but record counts vary by month and product. Government dominates segment counts; countries have equal counts. Company-level balance cannot be tested. Fiscal-year equivalence is not established.

There are no complete duplicates. The five-field combination Segment + Country + Product + Discount Band + Date yields 12 repeated occurrences; adding Sale Price produces a unique six-field combination. This price-dependent candidate needs business validation and should not be mistaken for a stable source identifier.

The file's strengths are explicit sales/profit-related fields, complete populated columns, and categorical and monthly context. Its limits include unknown provenance and units, no company identity, uncertain aggregation, formatted measures, short history, and absent obligation, budget, cash, and document records. Explicitly authorized temporary parsing shows right-skewed sales/cost/profit measures. Sales has a mean of 169,609.07 versus median 35,540.20; Profit ranges from -40,617.50 to 262,200.00, with 58 negative observations under the parentheses convention. Discounts statistics exclude 53 unresolved placeholders and Profit statistics exclude 5. These observations identify questions for later EDA, not errors or cleaning decisions.

Proceeding later to detailed EDA should investigate source documentation and entity scope; field definitions and currencies; formatting, placeholder, and sign conventions; record grain and stable identifiers; potential gross-sales/discount/sales/cost/profit relationships; repeated dimensional combinations; period completeness and sampling; and contextual distributions under source-validated numeric interpretations. Any future cleaning rules need evidence and separate approval. Current usefulness is descriptive financial analysis and potential anomaly exploration, not complete coverage of the V1.

# Dataset Comparison — Financial Manager AI

Usefulness depends on the capability and analytical question. Neither dataset is universally better. Receiver General figures below come from its unchanged analysis outputs, with documentation corrections noted in sections 2, 4, and 9; Company Financials findings come from the new outputs above.

| Characteristic | Receiver General | Company Financials |
|---|---|---|
| Record granularity | Voucher entries and accounting/control summaries are supported by documentation; exact local row types remain unclassified (section 4) | Provisional sales observation by segment/country/product/discount band/month; aggregation unconfirmed |
| Number of records | 9,282 | 700 |
| Number of attributes | 16 | 16 |
| Temporal coverage | Effective dates April 3, 2023–January 3, 2024; 188 distinct dates; fiscal labels require reconciliation | 16 consecutive monthly labels, September 2013–December 2014; 175 rows in 2013 and 525 in 2014 |
| Main identifiers | Accounting control, voucher item, GBS control; voucher ID mostly absent | No explicit record/company ID; empirical dimension-plus-price combination only |
| Main dimensions | Department, voucher type, control type, debit/credit, fiscal period, subledger | Segment, country, product, discount band, month, year |
| Main financial measures | Numeric journal item amount; summary line count is a control measure | Eight text-stored measures: quantity, manufacturing/sale prices, gross sales, discounts, sales, COGS, profit |
| Missing-data concerns | Voucher ID 98.08%, ledger code 97.94%, coding block 96.96%, subledger 23.55% missing | No Pandas nulls; 53 Discounts and 5 Profit dollar-dash placeholders with unresolved meaning |
| Candidate key | Accounting Control Number + Journal Voucher Item Identifier, unique in file | Segment + Country + Product + Discount Band + Date + Sale Price, unique in file but price/format dependent |
| Complete duplicate rows | 0 | 0 |
| Distribution evidence | Numeric amounts strongly skewed; median about 1,767.72 versus mean about 329.81 million | Categorical imbalances and right-skewed numeric sales/cost/profit profiles; Sales mean 169,609.07 versus median 35,540.20 under temporary parsing conventions |
| Main analytical value | Accounting-entry exploration, department/debit-credit analysis, potential anomaly work | Sales/cost/profit exploration across products and segments, potential anomaly work |
| Provenance and semantics | Official PSPC catalogue/dictionary identified; Kaggle retained as recorded redistribution, exact version unverified; currency unconfirmed | Source and dictionary not found locally; dollar symbol does not identify currency |

## V1 Capability Coverage Matrix

| Capability | Receiver General | Company Financials | Combined Coverage |
|---|---|---|---|
| Accounts Payable | Inadequate | Inadequate | Inadequate |
| Financial Data Analysis | Suitable | Suitable | Suitable |
| Budget Variance | Inadequate | Inadequate | Inadequate |
| Financial Document RAG | Inadequate | Inadequate | Inadequate |
| Financial Anomaly Detection | Potentially Suitable | Potentially Suitable | Potentially Suitable |
| Cash Flow Forecast | Partial | Partial | Partial |

Combined coverage means a **portfolio of separate analytical sources**, not joined records or summed totals. The pair broadens financial-analysis examples from journal entries to sales and profitability, and offers different anomaly-investigation settings. It does not supply the missing obligation, budget, document, or cash evidence, and it does not extend a single entity's history.

**Potential integration:** neither inventory establishes a shared entity identifier, record mapping, currency, or compatible accounting basis. The observed time ranges do not overlap. Direct integration is therefore **not established**. Common-looking dates or financial amounts are not join keys. Any later integration would need evidence of shared business scope and a validated mapping.

**Independent usefulness:** Receiver General can support accounting-entry questions; Company Financials can support sales/profit questions. Both can inform separate anomaly experiments. Neither file establishes alignment with the provisional **single-company, BRL** V1 scope. Suitability for descriptive analysis does not resolve these operational-fit questions.

## Data Gaps for Financial Manager AI V1

| Capability | Evidence already available | Essential data or evidence still missing |
|---|---|---|
| Accounts Payable | Accounting amounts and sales/cost labels; neither establishes obligations | Suppliers, purchase invoice/obligation identifiers and amounts, issue dates, due dates, payment status, outstanding balances, and linked payments/adjustments for the target company |
| Financial Data Analysis | Dates/periods and amounts; Receiver General departments/accounting dimensions; Company Financials sales/cost/profit and commercial dimensions | Confirmed target-company identity, currency/units, field definitions, accounting/recognition basis, aggregation grain, and source provenance/coverage. Existing dimensions and actual measures need validation, not invention; a coherent target-company source is still needed to establish operational fit |
| Budget Variance | Potential actual amounts and temporal dimensions; departments in Receiver General and sales dimensions in Company Financials | Budget/planned amounts, version/scenario, target entity, and a mapping to actuals with matching periods, categories/departments or products, units, and recognition basis. Neither file supplies a compatible budget |
| Financial Document RAG | Structured records only | A relevant corpus such as contracts, policies, or financial reports, with document identity, source, date/version, and retrievable text; no documentary corpus is supplied |
| Financial Anomaly Detection | Record populations and contextual dimensions; numeric accounting amounts and text sales measures | Verified measure semantics and comparable groups, entity scope and grain, and a defensible evaluation reference (for example reviewed cases). Labeled anomalies are absent but are not necessarily required for every exploratory method; broader history is needed if temporal/seasonal detection is chosen |
| Cash Flow Forecast | Accounting dates/amounts and monthly sales/profit context | Confirmed dated cash inflows/outflows, cash-account identifiers, opening/observed balances, linked settlement information, known future receipts/payables, and sufficient same-entity history at the forecast frequency. Sales/profit and debit/credit do not establish these cash facts |

## Documentation Review Outcome — 2026-09-26

This documentation-only review updates the earlier unresolved-question list. Analysis code and outputs remain unchanged.

- **Fiscal convention documented; local alignment unresolved:** the official [Receiver General fiscal-month table](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/paiements-payments/soutien-support-eng.html) maps 1–12 to April–March. This is a related Receiver General publication, not an explicit mapping in this dataset's dictionary. Section 9 distinguishes that cross-document interpretation from the unexplained local prior-year/effective-date relationship.
- **Uniform granularity conclusion corrected:** the [dictionary](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/SGBT-BCMS-dd.xml) defines grouped detail counts as well as voucher-entry identifiers, and the [dataset-specific guide](https://donnees-data.tpsgc-pwgsc.gc.ca/ba1/tcrg-rgat/soutien-support-eng.html) distinguishes voucher transactions and control data. Multiple record types are supported; their exact distribution and relationships in this file are unverified. Existing key tests remain valid for this extract only.
- **Authoritative source and dictionary resolved:** section 2 now links PSPC's [official catalogue](https://open.canada.ca/data/en/dataset/fa79f13c-2796-41a9-a9cd-28650ebe0f06), dictionary, and technical support page. Kaggle remains the recorded redistribution source, but its listing/version and the local extract's provenance chain remain unverified.

These findings justify documentation corrections, not changes to numeric outputs, dates, data-quality labels, or the capability matrix. No EDA or raw-data change was performed. The Company Financials inventory is unchanged.

## Recommended Next Data Step

1. **Retain both datasets as candidates for separately authorized detailed EDA.** Receiver General can be investigated for accounting semantics, aggregation, missingness, and contextual amounts. Company Financials needs source/field clarification, validation of the temporary numeric parsing conventions, and investigation of placeholder meanings and contextual distributions. Neither is accepted yet as the operational V1 dataset.
2. **Seek additional coherent target-company data for Accounts Payable, Budget Variance, Financial Document RAG, and Cash Flow Forecast.** Financial Data Analysis and exploratory anomaly work have candidate inputs, but single-company/BRL fit and validated semantics remain unresolved.
3. **The next acquisition target should be a documented, linked single-company financial dataset**, ideally in BRL, containing supplier invoices/obligations, due dates, outstanding balances, payments and dated cash movements with account balances. Matching budget versions and source documents would then fill the remaining gaps. This is a recommendation only; no search or download has been performed.
4. **Synthetic complementary data may eventually be necessary** if a coherent, accessible real source cannot cover the missing workflows. Any such records must be explicitly labeled synthetic and internally consistent; they would support demonstrations/tests, not establish real-world forecast or anomaly accuracy. No synthetic data is created now.
5. **Keep decisions open** on dataset selection and entity/currency scope; record grain and identifiers; recognition/sign/date conventions; obligation/payment/cash relationships; budget mappings and versions; anomaly comparison groups and evaluation; forecast horizon/frequency/history; document corpus; and later cleaning/schema choices until the required evidence is available.

**Exact next step:** review this inventory and decide whether to authorize detailed EDA on both files, beginning with source semantics, placeholder conventions, record grain, and contextual distributions, and a separate search for the linked single-company obligations/payments/cash dataset described above. Stop here: no cleaning, schema design, ML, RAG implementation, or application work is authorized by this inventory.